Feature/validation and minor improvements

Duslerke opened the pull request:
# What:
- Added missing tests and Rules for validator.
- Modified the names of controller tests so that they would reflect better what's actually being tested, rather than rely on developer on making hidden connections between elements within.
- Did minor changes within the tests to better match the name change, as well as the concept being tested.
- Modified logger tests and implementation to reflect changes done in terms of parameters controller needs to operate.
- Modified the initial logger call to display explicitly that the value is 'null' if it is. (Test comes first, of course)
- Added regions to the tests helper.

# Why:
 - Missing rules added so that whitespaces wouldn't get passed the validator, as well as would give a specific error message relating to the input.
- Modified tests' names so that the dependency of the arrange to the assert would be more clear for whoever reads these tests in the future be it new developer or be it whoever wrote this, but forgot what this does.
- Modified logger, to capture full request contents. Reason being, you typically look at the logs when something goes wrong. The endpoint behaviour depends on the newly added 2 extra fields to the request object, hence they should be logged to have full information when potential issue happens.
- Made the initial call also display the 'null', when value is null to make reading logs easier.
- Added regions to the tests helper as it was getting difficult to find the method in question.

# Notes:
- All changes done bellow are for GetProcessImage endpoint.
- More details on the reasoning behind changes can be found in the commit message details.

Duslerke committed:
Modified the GetProcessImageRequest Validator.

Added 2 extra tests for the 2 new fields of GetProcessImageRequest object ("processType, fileExtension). These tests are about throwing an error when a blank parameter is supplied. Added implementaion to validator based of these tests.

Also added back the removed tests and validation for the "processRef" and "imageId" fields. The ones that did the not null, not empty checks. Reason for it is that even though request with such null values wouldn't pass the validation, it wouldn't give the error message specific enough to what the issue is. I feel that providing 'null' argument versus an invalid one should return different error messages.

Now, there was an argument raised on the PR against these checks saying that: since now the parameters are being passed in through url path, when you don't provide them, it will simply be interpreted as a different path - hence 404 path not found.
However, in response to that, I'd say: if you had provided a url encoded whitespace (which sometimes gets encoded automatically) as an argument, it would count as a valid path, and the  .NET would interpret the whitespace as 'null'. I checked it through the browser and debugger - it confirms it.

Example: "{insert your host}/api/v1/processImageData/%20/a68e234a-bd08-3542-4b89-42572b6adc65/e68e234a-bd08-3542-4b89-42572b6adc65/%20"

Duslerke committed:
Added regions to MatProcessDataHelper for easier navigation.

Also changed class to static, since it contained nothing but static methods anyway. There's no need to be false hinting that you might need an instance of it.

Duslerke committed:
Renamed the GetProcessImage endpoint Controller tests to be more desciptive about what they actually test.

-------------- Request unimportant --------------

In a lot of cases where the test names starts with: "given_a_valid_request" or "given_an_invalid_request". The request object created or referenced in the name does not matter at all. Example:
Ln189 "given_a_valid_request_when_GetProcessImage_controller_method_is_called_then_it_returns_a_200_Ok_response"
this test name was misleading because due to this being a unit test - all the dependencies of class under test (in this case controller) are being mocked. The mock itself contains no logic apart from whatever you make it return when doing mock setup in the arrange step. So in this test's case it means that it does not matter what request I pass into it, it will always return whatever I specified that mock to return. So, if I'm testing "given_a_valid_request", what it actually means is that I'm testing whether whatever is specified happens based on request validation logic returning success. So the test is more like "under_the_validation_control_flow_of_success_X_happens".

To make this explicit, I renamed the test as well as moved empty object creation directly into argument place in the controller call (act phase). To signify that the request itself is not part of the test context, hence shouldn't be in the arrange step.

There are many more cases like this, so I won't cover them.

------------------Status Codes------------------

Ln263 status code wasn't even being used, because the test didn't even check for that - that was the job for the test before it. It's good to have these tests separated for easier and quicked troubleshooting.

Ln289 Since the test states that it's checking for 500 status code result in its name, this means that it should probably be put in the arrange step to make this more explicit.

--------------------"Correct Data"--------------------

I figured that it's a pretty bad description for whatever you're expecting as a result from the test. This is because this 'term' is too abstract and can mean too many things. So I changed this term into something more specific accross the Get endpoint tests in order to clarify what we're actually testing for.

----------------Modified few logger tests----------------

Tests in lines: "Ln423" -> "Ln411" and "Ln445" -> "Ln431" were indeed testing what they said, however they did not reflect the intention behind the tests.

The point of those tests was to show that logger is being called regardless of validation controll flow - as in it was called prior to the validation, however all they were testing was that the logger was being called at least once accross the entire validation controll flow.

So I changed these test to test exactly what they meant - that the request gets registered before it gets validated, meaning this registering action does not depend on the validation result.

To do this I had to ensure there's no Validator mock setup, which would influence the decision to call logger or not. This of course means a crash within the controller, since it's trying to make a decision based on null result now, but this crash can get mitigated by putting an empty try-catch block to ignore the error.

The whole idea is:
- If logger call depends on the validation result, due to a crash (and absense of control flow decision) in the controller, the call  to logger will never happen. And the test will fail due to there being 0 calls to logger in the assert step.
- However, if the logger call is independant of validation controll flow, then we should expect the call to happen prior to validation (prior to the error), and the test passes.

Now the counter argument could be: "what if the logger get called after all the validation, or in every branch of validation decision?". I'd say that even under these circumstances, you're still dependent on the validation - maybe not on the decision it makes, but on it not crashing, which is a bad way to ensure that every request coming into the controller gets registered.

Duslerke committed:
Modified GetProcessImage logger tests and logger.

Modified the tests to expect 4 properties logged from the GetProcessImageRequest object. (Previously it was only 2, only after that additional 2 were added.)

Added the implementation to make modified tests pass.

Modified couple names of the logger tests again to be more explicit that this logger call does not depend not only on the result of the validation, but the behavior (like crashing) as well.

Added additional test that ensures that the initial logger call will also have null values from the request replaced with "null" strings. (In order to have some consistency accross the calls; Also helps with troubleshooting )

LBHTKarki commented on mat-process-api/V1/Validators/GetProcessImageRequestValidator.cs:
```diff
@@ -13,10 +13,12 @@ public GetProcessImageRequestValidator()
+                .Must(ValidateGuid).WithMessage("You need to provide a valid process reference.");
+            RuleFor(req => req.imageId).NotNull().WithMessage("Image Id must be provided.").NotEmpty().WithMessage("Image Id must be provided.")
+                .Must(ValidateGuid).WithMessage("You need to provide a valid Image Id.");
+            RuleFor(req => req.processType).NotNull().WithMessage("Process Type must be provided.").NotEmpty().WithMessage("Process Type must be provided.");
```
would it be worth specifying in the NotEmpty message that the value cannot be empty? Or if same message is used then combining those two checks i.e. NotNull().NotEmpty().WithMessage("...

Duslerke replied:
Tried this `NotNull().NotEmpty().WithMessage("...`  before it didn't return the message on notnull check.
But I agree, I was thinking the exact same thing, it would be nice to have different error messages. I'll need to split the tests to do this though.

Duslerke replied:
Change made.

MirelaGeorgiou commented on mat-process-api.Tests/V1/Controllers/ProcessDataImagesControllerTests.cs:
```diff
@@ -301,21 +296,20 @@ public void given_an_unexpected_error_would_be_thrown_when_GetProcessImage_contr
 
         [Test]
-        public void given_an_unexpected_error_would_be_thrown_when_GetProcessImage_controller_method_is_called_then_it_returns_a_correct_error_message()
+        public void given_an_unexpected_error_would_be_thrown_when_GetProcessImage_controller_method_is_called_then_it_returns_a_correctly_formatted_error_message_that_is_based_of_exception()
```
You shouldn't use 'Returns' and generate an error. Use .Throws<> instead for when mocking the use case in this test.

Duslerke replied:
I wasn't happy with that implementation either. I'll try and see if I can make throws work with exceptions helper. 

But reviewing your PR gave me an idea of rewriting helper to return errors rather than throw them. Somehow it never registered to me that you can put an exception into a variable :D

Duslerke replied:
Alright, I managed to get this working. Tests are much much simpler to read and follow now, also the helper method itself has become simpler. Thanks for the suggestion!

MirelaGeorgiou commented on mat-process-api.Tests/V1/Controllers/ProcessDataImagesControllerTests.cs:
```diff
@@ -420,16 +408,34 @@ public void when_GetProcessImage_controller_method_is_called_then_it_calls_the_l
 
             //act
-            _processImageController.GetProcessImage(request);
+            try { _processImageController.GetProcessImage(request); } catch (Exception ex) { } //the empty try-catch is needed to ignore the exception being thrown due to absense of validation result returned later down the line. The idea is that logger is the first thing that is being called once request comes in.
```
You should try to mock the validation result, rather than using a try/catch within a test.

Duslerke replied:
Not mocking the validation result is the entire idea behind this test. I explained it in the comment next to the test, and in the commit message in bit more detail.
If I mock the result then I'll change what the test is asserting.

MirelaGeorgiou replied:
Please correct me if I am wrong, but the test aims at verifying that the logger will log the first message, even if the values passed are null? The logger should be called after the logging message, which means that the message should be logged irrespective to the validation result (positive or not). If that is the case, you can still mock the validation result and test that the logger logged the first line.

Duslerke replied:
The test aims to show that the request gets logged at least once regardless of validator behaviour.

By behaviour I mean, it shouldn't matter whether the validator returned true or false, or if it crashed entirely.

I think that if your initial log is dependent on the validator not crashing, then it's a bad of way of logging that (and what) request happened.

So I set up the test in such a way, that doesn't have validator setup to show that despite the validator crash (which happened in the try-catch block), the logger was indeed called at least once. 

This also demonstrates pretty clearly that whatever happens in the validator has no effect on the request getting logged.

Duslerke replied:
I'll update the comments next to the test to make this clearer. I can see where the confusion is coming from now that I read it again. 

Sorry, my bad. initially I thought it was pretty good comment, but forgot to clarify the bit about crashing the flow being necessary - only did that in the commit message.

Duslerke replied:
Added better description now above the 2 tests.

MirelaGeorgiou commented on mat-process-api.Tests/V1/Validators/GetProcessImageRequestValidatorTests.cs:
```diff
@@ -20,6 +20,62 @@ public void SetUp()
+        [TestCase(null)]
+        [TestCase("")]
+        [TestCase(" ")]
+        public void given_a_request_with_null_empty_or_whitespace_processRef_when_getProcessImageRequestValidator_is_called_then_it_returns_an_error(string processRef)
```
As discussed earlier, null should not be tested as it will fire a 404 error. It's empty/whitespace possibility that we should validate against.

Duslerke replied:
It won't fire a 404 error because the .NET translates the whitespace into null within a request object.
Here is the screenshot of debugging the request like that:

![image](https://user-images.githubusercontent.com/43747286/72914781-2155fb00-3d37-11ea-8cf9-d88b08439526.png)

Duslerke committed:
Modified GetProcessRequestValidator, made it return different error messages.

The validator now returns different error messages based on whether the field was null or empty. Pretty handy for whoever is consuming or trying to connect to the API.

Duslerke committed:
Modified ErrorThrowerHelper to return Exceptions rahter than throw them.

Motivation behind it is easier to read tests, as you no longer have to do the .Returns(()=> ...) method to throw an exception. Also no longer need the pointless return within that same Return to keep compiler happy.

Another bonus is that tests setup is much simpler looking than before making test easier to read and understand.

The helper method also benefits from this by not needing parameters anymore. The sole reason it had parameters in before was so that the same error could be thrown twice. Now it's done, by returning random error, and seting up its throw.

Duslerke committed:
Improved ErrorThrowerHelper dual-layer base error messages.

So that they would be easier to distinguish if debugger use is needed.

Duslerke committed:
Clarified the GetProcessImage endpoint initial logger call test.

Had to use summary not to get limited by trying to make single line comment at sane levels of length.

Had to clarify this bit better, due to previous explanation not being clear enough, and bit confusing in regards to "making validation crash" bit.
